pkg/agent/flowexporter/connections/conntrack_connections.go

```diff
@@ -137,18 +114,52 @@ func (cs *ConntrackConnectionStore) Poll() ([]int, error) {
 		}
 	}
 	var totalConns int
+	var filteredConnsList []*flowexporter.Connection
 	for _, zone := range zones {
-		filteredConnsList, totalConnsPerZone, err := cs.connDumper.DumpFlows(zone)
+		filteredConnsListPerZone, totalConnsPerZone, err := cs.connDumper.DumpFlows(zone)
 		if err != nil {
 			return []int{}, err
 		}
 		totalConns += totalConnsPerZone
-		// Update only the Connection store. IPFIX records are generated based on Connection store.
-		for _, conn := range filteredConnsList {
-			cs.AddOrUpdateConn(conn)
-		}
+		filteredConnsList = append(filteredConnsList, filteredConnsListPerZone...)
 		connsLens = append(connsLens, len(filteredConnsList))
 	}
+
+	// Reset IsPresent flag for all connections in connection map before updating
+	// the dumped flows information in connection map. If the connection does not
+	// exist in conntrack table and has been exported, then we will delete it from
+	// connection map. In addition, if the connection was not exported for a specific
+	// time period, then we consider it to be stale and delete it.
+	deleteIfStaleOrResetConn := func(key flowexporter.ConnectionKey, conn *flowexporter.Connection) error {
+		if !conn.IsPresent {
+			// Delete the connection if it is ready to delete or it was not exported
+			// in the time period as specified by the stale connection timeout.
+			if conn.ReadyToDelete || time.Since(conn.LastExportTime) >= cs.staleConnectionTimeout {
+				if err := cs.deleteConnWithoutLock(key); err != nil {
+					return err
+				}
+			}
+		} else {
+			conn.IsPresent = false
+		}
+		return nil
+	}
+
+	// Hold the lock until we verify whether the connection exist in conntrack table,
+	// and finish updating the connection store.
+	cs.AcquireConnStoreLock()
+
+	if err := cs.ForAllConnectionsDoWithoutLock(deleteIfStaleOrResetConn); err != nil {
+		return []int{}, err
```

Comment:
shouldn't the lock be released here?

Reply:
Previously the lock is released here, and it caused a race condition error. `deleteIfStaleOrResetConn` and `AddOrUpdateConn`(L158) both grab and release the lock separately, and they both modify `conn.IsPresent` field.  Ideally the value of `conn.IsPresent` should only be released until both two functions finish their execution. The error happens when Flow exporter grab the lock between `deleteIfStaleOrResetConn` and `AddOrUpdateConn`, and read a wrong value of `conn.IsPresent`. 

Please refer to the PR description for more detailed explanation.

Reply:
I meant release the lock inside the error clause if `err != nil` (sorry the line of comment should have been 152)

Reply:
updated. Thanks.